Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/SNNClusterer.java

```diff
@@ -0,0 +1,637 @@
+        this.clusters = new ArrayList<>();
+        this.clustered = new HashSet<>();
+        this.visited = new HashSet<>();
+        this.currentCluster = new ArrayList<>();
+        this.noiseCluster = new ArrayList<>();
+        this.pointIndex = 0;
+
+        // Noise cluster will be 0th cluster, start core clusters index at 1
+        this.clusters.add(0, this.creator.createCluster(this.noiseCluster));
+        this.clusterCount = 1;
+
+        // Ready to learn.
+        return true;
+    }
+
+    /**
+     * Processes the next point by checking if the point has been visited and if
+     * not, calling the method to determine the number of neighbors with at
+     * least eps shared neighbors. If this is larger than min points, the expand
+     * cluster method is called. Otherwise the point is added to the noise
+     * cluster.
+     *
+     * @return boolean representing whether there are more points left to
+     * process.
+     */
+    @Override
+    protected boolean step()
+    {
+        // Retrieve the point to process
+        DataType point = this.points.get(this.pointIndex);
+        if (!this.visited.contains(point))
+        {
+            // Mark point as visited
+            this.visited.add(point);
+
+            // Get neighbors with at least eps common neighbors
+            List<DataType> neighbors = this.getConnectedNeighbors(point);
+
+            // If number of neighbors with at least eps shared neighbors is below 
+            // the min points threshold, add point to noise cluster
+            if (neighbors.size() < this.minPts)
+            {
+                this.noiseCluster.add(point);
+                this.clustered.add(point);
+            }
+            // If there are at least minpts neighbors with at least eps shared 
+            // neighbors consider the point to be a core point
+            else
+            {
+                // Expand Cluster around the core point
+                this.expandCluster(point, neighbors);
+
+                // Add expanded cluster to list of clusters
+                this.clusters.add(this.clusterCount, this.creator.createCluster(
+                    this.currentCluster));
+
+                // Increase cluster count and make new empty current cluster
+                this.clusterCount++;
+                this.currentCluster = new ArrayList<>();
+            }
+        }
+
+        // Increase the point index and determine whether there are more points 
+        // to process
+        this.pointIndex++;
+        return (this.pointIndex < this.points.size());
+    }
+
+    /**
+     * Returns a list of all the points with at least eps shared nearest
+     * neighbors.
+     *
+     * @param point The point to get the neighbors list for.
+     * @return The list of all points with >= eps shared neighbors
+     */
+    private List<DataType> getConnectedNeighbors(DataType point)
+    {
+        List<DataType> result = new ArrayList<>();

```

Comment:
For performance you may want to initialize this to the expected number of neighbors (you can trim it at the end)